PowerEditor/src/MISC/Common/Common.cpp

```diff
@@ -1158,38 +1158,84 @@ bool isCertificateValidated(const generic_string & fullFilePath, const generic_s
 		errorMessage += GetLastErrorAsString(GetLastError());
 		MessageBox(NULL, errorMessage.c_str(), TEXT("Certificate checking"), MB_OK);
 	}
-
-	// Clean up.
-	if (pSignerInfo != NULL) LocalFree(pSignerInfo);
-	if (pCertContext != NULL) CertFreeCertificateContext(pCertContext);
-	if (hStore != NULL) CertCloseStore(hStore, 0);
-	if (hMsg != NULL) CryptMsgClose(hMsg);
-	if (szName != NULL) LocalFree(szName);
-
-	return isOK;
-}
-
-bool isAssoCommandExisting(LPCTSTR FullPathName)
-{
-	bool isAssoCommandExisting = false;
-
-	bool isFileExisting = PathFileExists(FullPathName) != FALSE;
-
-	if (isFileExisting)
-	{
-		PTSTR ext = PathFindExtension(FullPathName);
-
-		HRESULT hres;
-		wchar_t buffer[MAX_PATH] = TEXT("");
-		DWORD bufferLen = MAX_PATH;
-
-		// check if association exist
-		hres = AssocQueryString(ASSOCF_VERIFY|ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_COMMAND, ext, NULL, buffer, &bufferLen);
-        
-        isAssoCommandExisting = (hres == S_OK)                  // check if association exist and no error
-			&& (buffer != NULL)                                 // check if buffer is not NULL
-			&& (wcsstr(buffer, TEXT("notepad++.exe")) == NULL); // check association with notepad++
-        
-	}
-	return isAssoCommandExisting;
-}
\ No newline at end of file
+
+	// Clean up.
+	if (pSignerInfo != NULL) LocalFree(pSignerInfo);
+	if (pCertContext != NULL) CertFreeCertificateContext(pCertContext);
+	if (hStore != NULL) CertCloseStore(hStore, 0);
+	if (hMsg != NULL) CryptMsgClose(hMsg);
+	if (szName != NULL) LocalFree(szName);
+
+	return isOK;
+}
+
+bool isAssoCommandExisting(LPCTSTR FullPathName)
+{
+	bool isAssoCommandExisting = false;
+
+	bool isFileExisting = PathFileExists(FullPathName) != FALSE;
+
+	if (isFileExisting)
+	{
+		PTSTR ext = PathFindExtension(FullPathName);
+
+		HRESULT hres;
+		wchar_t buffer[MAX_PATH] = TEXT("");
+		DWORD bufferLen = MAX_PATH;
+
+		// check if association exist
+		hres = AssocQueryString(ASSOCF_VERIFY|ASSOCF_INIT_IGNOREUNKNOWN, ASSOCSTR_COMMAND, ext, NULL, buffer, &bufferLen);
+        
+        isAssoCommandExisting = (hres == S_OK)                  // check if association exist and no error
+			&& (buffer != NULL)                                 // check if buffer is not NULL
+			&& (wcsstr(buffer, TEXT("notepad++.exe")) == NULL); // check association with notepad++
+        
+	}
+	return isAssoCommandExisting;
+}
+
+#ifndef _WIN64
+static ULONGLONG filetime_to_time_ull(const FILETIME* ft)
+{
+    ULARGE_INTEGER ull;
+    ull.LowPart = ft->dwLowDateTime;
+    ull.HighPart = ft->dwHighDateTime;
+    return (ull.QuadPart / 10000000ULL - 11644473600ULL);
+}
+
+int winapi_wstat(wchar_t const* _FileName, struct _stat* _Stat)
+{
+    // Starting from Visual Studio 2015, _wstat always returns -1 in Windows XP.
+    // So here is a WinAPI-based implementation of _wstat.
+    int nResult = -1;
+    DWORD dwAttr = ::GetFileAttributes(_FileName);
+    if (dwAttr != INVALID_FILE_ATTRIBUTES)

```

Comment:
Since you're checking for CreateFile return value, this check is redundant.

Reply:
The intention was: if GetFileAttributes returns INVALID_FILE_ATTRIBUTES, we don't even call CreateFile. These are two different functions, so checking results of both sounds logical, no?

Reply:
Sure, but you're not winning anything with this check. If the file/directory does not exist, `CreateFile` will fail anyway. Right now you are checking for file existence twice, and if `GetFileAttributes` returns valid data you still can't be sure that the file will still exist when `CreateFile` gets called (may very well be deleted between those two calls).